---
title: "ATACFlow-Introduction"
author: "Zheng Wei and Wei Zhang"
date: "`r Sys.Date()`"
output: html_document
vignette: >
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteIndexEntry{An Introduction to ATACFlow}
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Introduction

The ATACFlow package provides a dataflow graphs organized pipeline 
for quantifying and annotating ATAC-seq and DNase-seq Reads in R,
which integrate the functionality of several R packages 
(such as Rsamtools, ChIPpeakAnno and so on) and external softwares 
(e.g. [AdapterRemoval](https://github.com/MikkelSchubert/adapterremoval)[1], [bowtie2](http://bowtie-bio.sourceforge.net/bowtie2/index.shtml)[2], 
through the Rowtie2 package and 
[Fseq](http://fureylab.web.unc.edu/software/fseq)[3]). 
Users can call the preset pipeline functions or 
rebuild the workflow with elements process of ATACFlow
for sequence data (FASTQ, SAM, BAM, BED file) complete/partial processing 
and analysing easily and flexibly in a single R script.
That will be convenient to migrate, share and reproduce all 
details such as parameters settings, intermediate result and so on.
Besides, a quality control report file in HTML, 
which is able to be viewed in web browser,
will be created in preset pipelines.

ATACFlow can be easily installed on various operator system platforms
(Windows, Linux, MacOs). All functions in package comsume up to 16G memery.
Most function only consume less than 8G. So the package is available for 
not only servers but also  most of PC.
ATACFlow supports analysis of both single end reads and
paired-end ATAC-seq and DNase-seq sequencing reads data.
We have successfully apply the pipeline on 
raw datasets (FASTQ files) from GEO. 
The related sequencing platform are Illumina.
All valid standard format intermediate result files
(FASTQ, SAM, BAM, BED file) generated
by other program (such as BAM BED files from ENCODE) 
are also tested by rebuilt subpipeline.


This package is developed and maintained by members of 
[Xiaowo Wang Lab](http://bioinfo.au.tsinghua.edu.cn/member/xwwang)

MOE Key Laboratory of Bioinformatics and Bioinformatics Division, 

TNLIST / Department of Automation, Tsinghua University

contact:{wei-z14,w-zhang16}(at)mails.tsinghua.edu.cn

## Preparation

### Package Installation

To install the latest version of ATACFlow, you will need to be using the
latest version of R. ATACFlow is part of Bioconductor project, 
so you can install ATACFlow and its dependencies like this:

```{r install, eval=FALSE}
source("http://www.bioconductor.org/biocLite.R")
biocLite("ATACFlow")
```

### Loading

Just like other R package, 
you need to load ATACFlow like this each time before using the package.
If you need to use fseq, 
we recommed to set max memory size for java (8G, 8000M in the example).
Or the rJava will use the default parameter for fseq.

```{r loading,message=FALSE}
options(java.parameters = "-Xmx8000m")
library(ATACFlow)
```

### Loading Recommend Package

The BSgenome package, TxDb known gene package and OrgDb annotation package
for some functions are requied. The users had better install and
load the specific species related packages before using the packages.

```{r loadingpkg,message=FALSE}
library(magrittr)
library(BSgenome.Hsapiens.UCSC.hg19)
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
library(org.Hs.eg.db)
library(R.utils)
```


### Configure Other Parameters

These configurations are also optional. 
"tmpdir" is the path to save all of the temporary data and
the default result storage path. If it is configured,
current work directory will be set as "tmpdir".
"threads" is the maximum threads allowed to 
be created for data processing.
The default value is 1.


```{r loadref, eval=FALSE}
setConfigure("tmpdir","path/to/tmpdatafolder")
setConfigure("threads",8)
```

### Reference Data Installation and Loading

We strongly recommend to install referece data first before using the package
although it is optional. 
"refdir" is the folder that will save all of the reference data.
"genome" is the genome name like hg19, hg38, mm10, mm9 and so on.
The program will detect the elements that have not been installed and install them.
Some resources need to be downloaded from internet. 
So don't forget to connect internet during installation.
Or the installation will be failed.
If all of the refernce data was installed, 
these two lines still need to be called 
for configurating the reference data path and genome. 

```{r config, eval=FALSE}
setConfigure("refdir","path/to/refdatafolder")
setConfigure("genome","hg19")
```

The installation will consume several hours for data download,
building bowtie2 index depending on computer performance and 
network bandwidth. 

If the reference data is not configured, 
the related argument of functions has to be set manually during using. 


## Preset Pipeline

Note: The users have to configure step by step in last section to call preset pipeline.
All of the reference, temperary data storage path, max memery for rJava
and threads allowd to be created will according to the configurations.


### case study

For single-end case study, a fastq file of sequencing data and 
its adpater are required. The preset pipeline can be
called like this:

```{r casesingle, eval=FALSE}
bedbzfile11 <- system.file(package="ATACFlow", "extdata", "chr20_1.fq.bz2")
bedbzfile12 <- system.file(package="ATACFlow", "extdata", "chr20_2.fq.bz2")
atacPipe(fastqInput1 = bedbzfile11,adapter1 = "CTGTCTCTTATACACATCTCCGAGCCCACGAGACTGAAG")
```

For paired-end case study, two fastq files of sequencing data are requied. 
The preset pipeline can be called like this:

```{r casepair, eval=FALSE}
bedbzfile11 <- system.file(package="ATACFlow", "extdata", "chr20_1.fq.bz2")
bedbzfile12 <- system.file(package="ATACFlow", "extdata", "chr20_2.fq.bz2")
atacPipe(fastqInput1 = bedbzfile11,fastqInput2 = bedbzfile12)
```

[Example report](http://bioinfo.au.tsinhua.edu.cn/member/zwei/example/Report.html)

### Case-Control Study

For case-Control study, paired-end sequencing data are supported only.
Two fastq files of sequencing data for case and control each are requied.
The preset pipeline can be called like this:

```{r casecontrol, eval=FALSE}
td <- tempdir()
bedbzfile11 <- system.file(package="ATACFlow", "extdata", "chr20_1.fq.bz2")
bedbzfile12 <- system.file(package="ATACFlow", "extdata", "chr20_2.fq.bz2")
bedbzfile21 <- system.file(package="ATACFlow", "extdata", "chr21_1.fq.bz2")
bedbzfile22 <- system.file(package="ATACFlow", "extdata", "chr21_2.fq.bz2")
atacPipe2(case=list(fastqInput1 = bedbzfile11,fastqInput2 = bedbzfile12),
          control=list(fastqInput1 = bedbzfile21,fastqInput2 = bedbzfile22))
```

[Example report](http://bioinfo.au.tsinhua.edu.cn/member/zwei/example/Report2.html)

## Build User-defined pipeline or Use functions indivially

All sub-processes are available for recombine new whole pipeline or
subpipeline easily and flexibly. They are also able to be called indivially.

### Preprocess

Users can use %>% to build a pipeline to obtain merged, renamed and
adapter removed clean reads fastq file(s) that is ready for mapping.


```{r Preprocess}
library(magrittr)
td <- tempdir()
setConfigure("tmpdir",td)

# Identify adapters
prefix<-system.file(package="ATACFlow", "extdata", "uzmg")
(reads_1 <-file.path(prefix,"m1",dir(file.path(prefix,"m1"))))
(reads_2 <-file.path(prefix,"m2",dir(file.path(prefix,"m2"))))

reads_merged_1 <- file.path(td,"reads1.fastq")
reads_merged_2 <- file.path(td,"reads2.fastq")
atacproc <- 
atacUnzipAndMerge(fastqInput1 = reads_1,fastqInput2 = reads_2) %>%
atacRenamer %>% atacRemoveAdapter

 
dir(td) 
```

### Mapping

If the reference has not been configured, the bowtie2 index should be built first.
Then bowtie2 mapping functions could used to map reads to reference genome.

```{r Mapping}
td <- tempdir()
setConfigure("tmpdir",td)

## Building a bowtie2 index
library("Rbowtie2")
refs <- dir(system.file(package="ATACFlow", "extdata", "bt2","refs"),
full=TRUE)
bowtie2_build(references=refs, bt2Index=file.path(td, "lambda_virus"),
"--threads 4 --quiet",overwrite=TRUE)
## Alignments
reads_1 <- system.file(package="ATACFlow", "extdata", "bt2", "reads",
"reads_1.fastq")
reads_2 <- system.file(package="ATACFlow", "extdata", "bt2", "reads",
"reads_2.fastq")
if(file.exists(file.path(td, "lambda_virus.1.bt2"))){
    (atacBowtie2Mapping(NULL,bt2Idx = file.path(td, "lambda_virus"),
       samOutput = file.path(td, "result.sam"),
       fastqInput1=reads_1,fastqInput2=reads_2,threads=3))
    head(readLines(file.path(td, "result.sam")))
}
```
### Convert SAM File to BED File

The mapping results are stored in a SAM file. 
SamToBed functions can covert it into BED file.
During coverting, the operation like sorting, shifting, filting chromosome and so on
can also be setted to do in the meantime.

```{r Convert}
library(R.utils)
library(magrittr)
td <- tempdir()
setConfigure("tmpdir",td)

sambzfile <- system.file(package="ATACFlow", "extdata", "Example.sam.bz2")
samfile <- file.path(td,"Example.sam")
bunzip2(sambzfile,destname=samfile,overwrite=TRUE,remove=FALSE)
samToBed(samInput = samfile) %>%
getReportVal("report")
```

### Filtering reads and Calling

Filter the nucleosome free reads(<100bp) for peak caling.

```{r FiltCalling}
library(R.utils)
library(magrittr)
td <- tempdir()
setConfigure("tmpdir",td)

bedbzfile <- system.file(package="ATACFlow", "extdata", "chr20.50000.bed.bz2")
bedfile <- file.path(td,"chr20.50000.bed")
bunzip2(bedbzfile,destname=bedfile,overwrite=TRUE,remove=FALSE)

bedUtils(bedInput = bedfile,maxFregLen = 100, chrFilterList = NULL) %>%
atacPeakCalling

dir(td)
```

### ATAC-seq Peak Annotation
ATAC-seq peak locate at open chromatin regions. Annotating these peak could find whether they locate at functional regions(such as promoter and enhencer). Function "peakanno" use  function "annotatePeak" in package "ChIPseeker"" to annotate ATAC-seq peak, for more parameters and usage, [click here](https://bioconductor.org/packages/release/bioc/html/clusterProfiler.html).


```{r echo=TRUE, results='hide', message=FALSE}
p1bz <- system.file("extdata", "Example_peak1.bed.bz2", package="ATACFlow")
peak1_path <- as.vector(bunzip2(filename = p1bz,
destname = file.path(getwd(), "Example_peak1.bed"),
ext="bz2", FUN=bzfile, overwrite=TRUE, remove = FALSE))
AnnoInfo <- peakanno(peakInput = peak1_path, TxDb = TxDb.Hsapiens.UCSC.hg19.knownGene, annoDb = "org.Hs.eg.db")
```

The output contains a pie chart in pdf format like below.
```{r echo=FALSE, results='hide', message=FALSE}
library(ChIPseeker)
```
```{r eval=TRUE, echo=FALSE, warning=FALSE}
peakanno <- AnnoInfo$getReportVal("annoOutput.rds")
plotAnnoPie(x = peakanno)
```

The function also generate a file contains annotation for all peaks.
Below is a part of the output.
```{r eval=TRUE, echo=FALSE, warning=FALSE}
peakanno <- as.data.frame(peakanno)
colnames(peakanno)[1] <- "chromatin"
peakanno <- subset(peakanno, select=c("chromatin", "start", "end", 
                                      "annotation", "geneStart", "geneEnd", 
                                      "geneId", "distanceToTSS", "SYMBOL", 
                                      "GENENAME"))
knitr::kable(peakanno[1:10, ])
```

### Go Analysis

GO is ......(using clusterProfiler)

The function need gene Id as input. User could choose different GO(molecular function, biological process and cellular component) according to different parameter. 

```{r echo=TRUE, results='hide', message=FALSE}
## extract gene ID
library(clusterProfiler)
data(geneList)
geneId <- names(geneList)[1:100]
goAna <- goanalysis(gene = geneId, OrgDb = "org.Hs.eg.db", keytype = "ENTREZID", ont = "MF")
```

The output file contains the GO term sorted by p-value, 
Below is a part of the output.

```{r eval=TRUE, echo=FALSE, warning=FALSE}
go_path <- goAna$getReportVal("goOutput")
go_data <- read.table(file = go_path, header = TRUE, sep = "\t")
go_data <- subset(go_data, select = c("ID", "Description", "GeneRatio", "pvalue", "qvalue"))
knitr::kable(go_data[1:10, ])
```

## Query



## Reference



