RMotifScan <- R6::R6Class(
  classname = "RMotifScan",
  inherit = BaseProc,

  public = list(
    initialize = function(atacProc, Seq_file = NULL, Motif = NULL, min.score = NULL,
                          output_file = NULL, Position = NULL, MP_file = NULL, editable = FALSE){
      super$initialize("RMotifScan",editable,list(arg1=atacProc))

      # necessary parameters
      if(!is.null(atacProc)){
        print("Parameter atacProc is not using now! We will add more functions in the future!")
      }
      private$paramlist[["Seq_file"]] <- Seq_file
      private$paramlist[["Motif"]] <- Motif
      private$paramlist[["min.score"]] <- min.score
      private$paramlist[["Position"]] <- Position

      # unnecessary parameters
      if(is.null(output_file)){
        private$paramlist[["output_file"]] <- paste0(dirname(Seq_file), "/output", collapse = "")
      }else{
        private$paramlist[["output_file"]] <- output_file
      }
      if(Position && is.null(MP_file)){
        private$paramlist[["MP_file"]] <- paste0(dirname(private$paramlist[["output_file"]]),
                                                 "/motif.mp", collapse = "")
      }else{
        private$paramlist[["MP_file"]] <- MP_file
      }

      # parameter check
      private$paramValidation()
    } # initialization end

  ), # public end

  private = list(
    processing = function(){
      private$writeLog(paste0("processing file:"))
      private$writeLog(sprintf("Sequence file:%s", private$paramlist[["Seq_file"]]))
      private$writeLog(sprintf("search file destination:%s", private$paramlist[["output_file"]]))
      private$writeLog(sprintf("position file destination:%s", private$paramlist[["MP_file"]]))
      ref <- Biostrings::readDNAStringSet(private$paramlist[["Seq_file"]])
      pwm <- private$paramlist[["Motif"]]
      sitesetList <- TFBSTools::searchSeq(x = pwm, subject = ref, min.score = private$paramlist[["min.score"]])
      tmp <- as(sitesetList, "data.frame")
      write.table(x = tmp, file = private$paramlist[["output_file"]], row.names = FALSE,  quote = FALSE)
      if(private$paramlist[["Position"]]){
        seq_info <- as.data.frame(stringr::str_split_fixed(string = tmp$seqnames, pattern = "[:-]+", n = 3))
        colnames(seq_info) <- c("chr", "start", "end")
        seq_info$start <- as.numeric(as.vector(seq_info$start)) + as.numeric(as.vector(tmp$start)) - 1
        seq_info$end <- as.numeric(as.vector(seq_info$start)) + as.numeric(nchar(tmp$siteSeqs)) - 1
        seq_info$strand <- tmp$strand
        seq_info$motif <- tmp$TF
        write.table(x = seq_info, file = private$paramlist[["MP_file"]], row.names = FALSE,  quote = FALSE)
      }


    }, # processing end

    checkRequireParam = function(){
      if(is.null(private$paramlist[["Seq_file"]])){
        stop("Parameter Seq_file is required!")
      }
      if(is.null(private$paramlist[["Motif"]])){
        stop("Parameter Motif is required!")
      }
      if(is.null(private$paramlist[["Position"]])){
        stop("Parameter Position is required!")
      }
    }, # checkRequireParam end

    checkAllPath = function(){
      private$checkFileExist(private$paramlist[["Seq_file"]])
      private$checkPathExist(private$paramlist[["output_file"]])
    } # checkAllPath end

  ) # private end

) # R6 class end

#' Finding motif binding sites in DNA sequences.
#'
#' @param Seq_file DNA sequence file, fasta or fastq.
#' @param Motif Motif PWM matrix generated by TFBSTools. Must be a SiteSet Class or SiteSetList Class.
#' @param min.score The minimum score for the hit. Can be given an character string in the format
#' of "80%" or as a single absolute value between 0 and 1. When it is percentage value, it represents
#' the quantile between the minimal and the maximal possible value from the PWM.
#' @param output_file Output file path and name, default: Seq_file_path/output.
#' @param Position TRUE or FALSE. Whether generate accuracy motif position file.Default: FALSE.
#' ATTENTION: If your DNA sequences are not from function "DnaSeqCut", be sure Position = FALSE,
#' otherwise the program will report an error.
#' @param MP_file If Position = TRUE, the accuracy motif position will be writen in this file.
#' Default: output_file_path/motif.mp
MotifScan <- function(atacProc = NULL, Seq_file = NULL, Motif = NULL, min.score = "80%",
                      output_file = NULL, Position = FALSE, MP_file = NULL){
  tmp <- RMotifScan$new(atacProc, Seq_file, Motif, min.score, output_file, Position)
  tmp$process()
  return(tmp)
}
