RMotifScan <- R6::R6Class(
  classname = "RMotifScan",
  inherit = BaseProc,

  public = list(
    initialize = function(atacProc, Seq_file = NULL, Motif = NULL,
                          output_file = NULL, editable = FALSE){
      super$initialize("RMotifScan",editable,list(arg1=atacProc))

      # necessary parameters
      if(!is.null(atacProc)){
        print("Parameter atacProc is not using now! We will add more functions in the future!")
      }
      private$paramlist[["Seq_file"]] <- Seq_file
      private$paramlist[["Motif"]] <- Motif

      # unnecessary parameters
      if(is.null(output_file)){
        private$paramlist[["output_file"]] <- paste0(dirname(Seq_file), "/output", collapse = "")
      }else{
        private$paramlist[["output_file"]] <- output_file
      }

      # parameter check
      private$paramValidation()
    } # initialization end

  ), # public end

  private = list(
    processing = function(){
      private$writeLog(paste0("processing file:"))
      private$writeLog(sprintf("Sequence file:%s", private$paramlist[["Seq_file"]]))
      private$writeLog(sprintf("destination:%s", private$paramlist[["output_file"]]))
      ref <- Biostrings::readDNAStringSet(private$paramlist[["Seq_file"]])
      pwm <- TFBSTools::toPWM(private$paramlist[["Motif"]])
      sitesetList <- TFBSTools::searchSeq(pwm, ref)
      tmp <- as(sitesetList, "data.frame")
      write.table(tmp, private$paramlist[["output_file"]], row.names = FALSE,  quote = FALSE)
    }, # processing end

    checkRequireParam = function(){
      if(is.null(private$paramlist[["Seq_file"]])){
        stop("Parameter Seq_file is required!")
      }
      if(is.null(private$paramlist[["Motif"]])){
        stop("Parameter Motif is required!")
      }
    }, # checkRequireParam end

    checkAllPath = function(){
      private$checkFileExist(private$paramlist[["Seq_file"]])
      private$checkPathExist(private$paramlist[["output_file"]])
    } # checkAllPath end

  ) # private end

) # R6 class end

#' Finding motif binding sites in DNA sequences.
#'
#' @param Seq_file DNA sequence file, fasta or fastq.
#' @param Motif Motif PFM matrix generated by TFBSTools. Must be a SiteSet Class or SiteSetList Class.
#' @param output_file Output file path and name, default: Seq_file_path/output.
MotifScan <- function(atacProc = NULL, Seq_file = NULL,
                      Motif = NULL, output_file = NULL){
  tmp <- RMotifScan$new(atacProc, Seq_file, Motif, output_file)
  tmp$process()
  return(tmp)
}
